rgmanager/man/clulog.8
.TH "clulog" "8" "Jan 2005" "" "Red Hat Cluster Suite Internal Commands"
.SH "NAME"
clulog \- Log a message to the cluster and/or system logs
.SH "SYNOPSIS"
.B clulog 
.B \-s
.I severity
.B [\-l
.I priority_filter
.B ] 
.B [\-n
.I program_name
.B ] 
.B [\-p
.I pid
.B ] 
.B message
.SH "DESCRIPTION"
.PP 
The
.B clulog
command sends a message to syslogd(8).
.SH "OPTIONS"
.IP "\-s <severity>"
Log at the specified severity (0-7; 0=ALERT, 7=DEBUG).
.IP "\-l <priority_filter>"
Apply the specified filter (0-7) for the message.  By default,
.B clulog
uses the service manager's assigned log level.
.IP "\-n <program_name>"
Log as the specified program name.
.IP "\-p <pid>"
Log as the specified PID.
.SH "SEE ALSO"
syslog(2), syslogd(8)


ChangeLog
* Wed Mar 25 2026 resource-agents contributors
- release candidate 4.18.0 rc1
- ocf-shellfunc: add ocf_log_pipe
- ocft: fix failing tests in resource-agents v4.17.0
- Filesystem: improve shell trace (set -x) output
- Filesystem: modify the return code when a mismatch between mount state and configuration is detected in monitor operation
- Filesystem: new force_unmount=move option
- Filesystem: optionally report "xargs ps -f" even when killing many processes
- Filesystem: signal many processes in parallel
- Filesystem: dont create systemd drop-in file for tmpfs/overlayfs
- Filesystem: try umount immediately after signals are sent
- IPaddr2: made find_interface() output empty when no interface is found
- IPsrcaddr: fix grep expression, so it doesnt log "stray \ before white space" with newer versions of grep
- aws-vpc-move-ip: add awscli_timeout parameter
- db2: set reintegration flag when promotion is successful
- docker: improve image existence check (#2121)
- exportfs: fix grep error on stop
- findif.c: remove unused colonptr variable
- pgsql: use monitor_user for monitor-calls and use .pgpass when monitor_password is not specified
- podman-etcd: add -a option to crictl ps (#2112)
- podman-etcd: enhance etcd data backup with snapshots and retention
- podman-etcd: fix "Peer URLs already exists" in add_member_as_learner (#2136)
- podman-etcd: fix learner node attribute not set after etcdctl failure
- podman-etcd: fix to prevent learner from starting before cluster is ready (#2098)
- podman-etcd: hardened monitor/stop actions
- podman-etcd: improve error handling to support retry on start errors (#2105)
- podman-etcd: preserve standalone voter identity during restart
- podman-etcd: prevent last active member from leaving the etcd member list
- podman-etcd: remove test code (#2103)
- podman-etcd: removed unneeded ETCDCTL_API environment variable
- podman-etcd: sync environment variables with Pod manifest
- portblock: check correct binary during validate-all
- portblock: monitor needs to also check state file of inverse action (#2108)
- powervs-move-ip/powervs-subnet: fix error logging
- powervs-subnet: wait until IP is activated before running monitor-check
- send_arp.linux/tickle_tcp: better alpine compatibility (#2119)
- sfex_lib: dont discard 'const' qualifier

* Wed Nov 19 2025 resource-agents contributors
- stable release 4.17.0
- nginx: restore SELinux context for pid-file during validate-all action

* Wed Nov 12 2025 resource-agents contributors
- release candidate 4.17.0 rc1
- ibm-cloud-vpc-cr-vip: new resource agent (#2002)
- ibm-cloud-vpc-move-fip: new resource agent (#2008)
- podman-etcd: new resource agent (#2023)
- powervs-move-ip: new resource agent (#2072)
- ocf-shellfuncs: add missing variable in crm_mon_no_validation()
- ocf-shellfuncs: fixed syntax error in crm_mon_no_validation() added at a25f08cf98d784894df9c52960eff5ccef059393
- ocf-shellfuncs: remove extra sleep from curl_retry (#2058)
- ocf-shellfuncs: set SHELL to default shell if it's set to nologin
- findif.sh: fix to avoid duplicate route issues
- spec: make nfs-utils a weak dependency
- spec: update Requires paths for RHEL10/CentOS 10
- AWS agents: reuse IMDS token until it expires (issue #1990) (#1991)
- All agents: fix shellcheck errors on output redirection [SC2328]
- All agents: use detected grep -E/-F instead of egrep/fgrep when detected
- Filesystem: dont report warnings when creating a Filesystem resource
- Filesystem: fix to avoid getting the wrong block device when doing grep
- Filesystem: remove validate-all fstype check, as there isnt a reliable and portable way to check
- Filesystem: speed up get pids (#2084)
- IPaddr2: add link status DOWN/LOWERLAYERDOWN check
- IPaddr2: add support for network interfaces altname in the monitor operation (#2027)
- MailTo: add s-nail support for multiple recipients
- aws.sh/ocf-shellfuncs: add ability to refresh token if it's invalid
- aws.sh: chmod 600 $TOKEN_FILE, add get_instance_id() with DMI support, and use get_instance_id() in AWS agents
- awsvip: let user specify which interface to use, and make the parameter optional in aws-vpc-move-ip
- azure-events-az: change variable name allNodesStopped to allResourcesNowStopped (#2030)
- azure-events-az: update Deployment info section
- azure-events: backport relevant changes from azure-events-az
- crypt: add tang/clevis support and fix some issues
- crypt: fix crypt_type check, and usage when action has not been specified
- crypt: fixes to validate-action, and to avoid running as a cloned resource
- db2: add "skip_basic_sql_health_check" parameter to avoid failing on systems with high load
- db2: add "monitor_retries", "monitor_sleep", and "monitor_retry_all_errors" parameters to be able to avoid failing on first try
- db2: use reintegration flag to avoid race condition on cluster reintegration, and removed FAL, as it's no longer needed
- mariadb: add SSL/TLS-Support (#2045)
- mdraid: add 'auto' and multiple devices support to parameter md_dev
- nfsserver: add ability to set e.g. "pipefs-directory=/run/nfs/rpc_pipefs" in /etc/nfs.conf to avoid issues with non-clustered Kerberized mounts
- nfsserver: dont use init-scripts if systemd is running
- nginx: fix ls-redirection, mute non-errors in validate-all, and set unique intervals for monitor actions
- openstack-cinder-volume: fix detach not working during start-action after #2000
- openstack-cinder-volume: wait for volume to be available
- oracle: improve monpassword description
- portblock: add nftables and multi-state support
- portblock: add Promoted monitor op, validate-all checks, and add "method" and "status_check" parameters
- portblock: improve version detection of iptables
- storage-mon: check if daemon is already running during start-action
- storage-mon: replace dashes with underscores in functions
- storage-mon: wait until monitor confirms the startup pid according to the OCF resource specification
- storage_mon: fix handling of 4k block devices

cman/qdisk/clulog.c
/*
  Copyright Red Hat, Inc. 2002
  Copyright Mission Critical Linux, 2000

  This program is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the
  Free Software Foundation; either version 2, or (at your option) any
  later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; see the file COPYING.  If not, write to the
  Free Software Foundation, Inc.,  675 Mass Ave, Cambridge, 
  MA 02139, USA.
*/
/** @file
 * Library routines for communicating with the logging daemon.
 *
 *  Author: Jeff Moyer <moyer@missioncriticallinux.com>
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdarg.h>
#include <malloc.h>
#include <dirent.h>
#include <signal.h>
#include <sys/errno.h>
#include <sys/types.h>
#include <sys/file.h>
#include <sys/socket.h>
#include <ccs.h>
#define SYSLOG_NAMES
#include <sys/syslog.h>
#undef SYSLOG_NAMES

#include <sys/wait.h>
#include <sys/types.h>
#include <linux/unistd.h>
#include <pthread.h>
#include <gettid.h>
#include <clulog.h>
#include <string.h>


#ifdef DEBUG
#include <assert.h>
#define Dprintf(fmt,args...) printf(fmt,##args)
#define DBG_ASSERT(x)  assert(x)
#else
#define Dprintf(fmt,args...)
#define DBG_ASSERT(x)
#endif

/*
 * Globals
 */
static int   log_is_open = 0;
static int   useconsole = 0;
static int   loglevel = LOGLEVEL_DFLT;
static int   syslog_facility = LOG_DAEMON;
static char  *daemon_name = NULL;
static pid_t daemon_pid = -1;
static pthread_mutex_t log_mutex = PTHREAD_MUTEX_INITIALIZER;

CODE logger_prioritynames[] = 
{ {"emerg", LOG_EMERG},
  {"alert", LOG_ALERT},
  {"crit", LOG_CRIT},
  {"err", LOG_ERR},
  {"warning", LOG_WARNING},
  {"notice", LOG_NOTICE},
  {"info", LOG_INFO},
  {"debug", LOG_DEBUG}
};

/*
 *  Exported Functions.
 */

/**
 * @return The current cluster log level.
 */
int
clu_get_loglevel(void)
{
	return loglevel;
}


/**
 * Set the cluster log level.
 *
 * @param severity	New log level.
 * @return 		Old log level, or -1 if 'severity' is an invalid log
 *			level.
 */
int
clu_set_loglevel(int severity)
{
	int ret = loglevel;

	if (severity > 0) {
		loglevel = severity;
		return ret;
	}

	return -1;
}


/**
 * @return The current cluster log facility.
 */
char *
clu_get_facility(void)
{
	int x = 0;

	pthread_mutex_lock(&log_mutex);
	for (; facilitynames[x].c_name; x++) {
		if (syslog_facility == facilitynames[x].c_val) {
			pthread_mutex_unlock(&log_mutex);
			return facilitynames[x].c_name;
		}
	}
	
	pthread_mutex_unlock(&log_mutex);
	return "daemon";
}


/**
 * Set the cluster log facility.
 *
 * @param facilityname  New log facility (see /usr/include/sys/syslog.h).
 * @return 		0
 */
int
clu_set_facility(char *facilityname)
{
	int x = 0, old;

	pthread_mutex_lock(&log_mutex);
	old = syslog_facility;

	for (; facilitynames[x].c_name; x++) {
		if (strcmp(facilityname, facilitynames[x].c_name))
			continue;
		syslog_facility = facilitynames[x].c_val;
		break;
	}

	if (syslog_facility == old) {
		pthread_mutex_unlock(&log_mutex);
		return 0;
	}

	closelog();
	log_is_open = 0;
	pthread_mutex_unlock(&log_mutex);
	return 0;
}


/**
 * Set the console logging mode.  Does not work for daemons.
 *
 * @param onoff		0 = off, otherwise on.
 * @return		Old log-to-console state.
 */
int
clu_log_console(int onoff)
{
	int ret = useconsole;

	useconsole = !!onoff;
	return ret;
}


/**
 * Cluster logging function.  Talks to syslog and writes to the
 * console, if necessary.
 */
int
do_clulog(int        severity,
	  int        write_to_cons,
	  pid_t      pid,
	  char       *prog,
	  const char *fmt, ...)
{
	va_list      args;
	char         logmsg[MAX_LOGMSG_LEN];	/* message to go to the log */
	char         printmsg[MAX_LOGMSG_LEN];	/* message to go to stdout */
	int          syslog_flags = LOG_NDELAY;

	pthread_mutex_lock(&log_mutex);
	if (severity > loglevel) {
		pthread_mutex_unlock(&log_mutex);
		return 0;
	}

	memset(logmsg, 0, MAX_LOGMSG_LEN);
	memset(printmsg, 0, MAX_LOGMSG_LEN);

	/*
	 * Check to see if the caller has forked.
	 */
	if (!pid) {

		/* Use thread IDs */
		if (daemon_pid != gettid()) {

			daemon_pid = gettid();
			log_is_open = 0;
		}

		syslog_flags |= LOG_PID;

	} else {

		daemon_pid = pid;
		closelog();
		log_is_open = 0;
		snprintf(logmsg, MAX_LOGMSG_LEN, "[%d]: ", pid);
	}

	if (prog) {

		if (daemon_name) {

			free(daemon_name);
			daemon_name = NULL;
		}

		daemon_name = strdup(prog);
	}

	if (!log_is_open) {

		openlog(daemon_name, syslog_flags, syslog_facility);
		log_is_open = 1;
	}
	/*
	 * Note: This can be called in the context of a CGI program, in which
	 * case anything printed to stdout goes to the web page.  This can
	 * cause problems if we have our standard <warning> strings b/c
	 * the web client will try to interpret this as an html tag.
	 */
	snprintf(logmsg + strlen(logmsg), MAX_LOGMSG_LEN - strlen(logmsg), 
		 "<%s> ", logger_prioritynames[severity].c_name);

	va_start(args, fmt);
	vsnprintf(logmsg + strlen(logmsg), MAX_LOGMSG_LEN - strlen(logmsg), 
		  fmt, args);
	va_end(args);

	if (write_to_cons || useconsole) {
		snprintf(printmsg, MAX_LOGMSG_LEN, "[%d] %s: ", daemon_pid,
			 logger_prioritynames[severity].c_name);

		va_start(args, fmt);
		vsnprintf(printmsg + strlen(printmsg),
			  MAX_LOGMSG_LEN - strlen(printmsg), fmt, args);
		va_end(args);

		fprintf(stdout, "%s", printmsg);
	}

	syslog(severity, "%s", logmsg);

	pthread_mutex_unlock(&log_mutex);

	return 0;
}


/**
 * Stop the cluster logging facility.
 */
void
clulog_close(void)
{
	closelog();
}
